src/containers/TransactionPage/TransactionPage.duck.js

```diff
@@ -390,12 +431,100 @@ export const declineSale = id => (dispatch, getState, sdk) => {
+  if (cancelOrInProgress(getState())) {
+    return Promise.reject(new Error('Cancel already in progress'));
+  }
+  dispatch(cancelSaleRequest());
+
+  const currentTransactions = await sdk.transactions.show({ id });
+  const transType = currentTransactions.data.data.attributes.lastTransition === TRANSITION_EXPIRE_FULL_REFUND_PERIOD
+    ? TRANSITION_CUSTOMER_CANCEL_NO_REFUND
+    : TRANSITION_CUSTOMER_CANCEL;
+
+  return sdk.transactions
+    .transition({ id, transition: transType, params: {} }, { expand: true })
+    .then(response => {
+      dispatch(addMarketplaceEntities(response));
+      dispatch(cancelSaleSuccess());
+      dispatch(fetchExpireCurrentUserNotifications());
+      return response;
+    })
+    .catch(e => {
+      dispatch(cancelSaleError(storableError(e)));
+      log.error(e, 'cancel-sale-failed', {
+        txId: id,
+        transition: transType,
+      });
+      throw e;
+    });
+};
+
+export const cancelSaleAfterAcceptedByCustomer = id => async (dispatch, getState, sdk) => {
+  if (cancelOrInProgress(getState())) {
+    return Promise.reject(new Error('Cancel already in progress'));
+  }
+  dispatch(cancelSaleRequest());
+
+  const currentTransactions = await sdk.transactions.show({ id });
+  const transType = currentTransactions.data.data.attributes.lastTransition === TRANSITION_ACCEPT
+    ? TRANSITION_CUSTOMER_CANCEL_BEFORE_EXPIRE
+    : TRANSITION_CUSTOMER_CANCEL_AFTER_EXPIRE;
+
+  return sdk.transactions
+    .transition({ id, transition: transType, params: {} }, { expand: true })
+    .then(response => {
+      dispatch(addMarketplaceEntities(response));
+      dispatch(cancelSaleSuccess());
+      dispatch(fetchExpireCurrentUserNotifications());
+      return response;
+    })
+    .catch(e => {
+      dispatch(cancelSaleError(storableError(e)));
+      log.error(e, 'customer-cancel-sale-failed', {
+        txId: id,
+        transition: transType,
       });
       throw e;
     });
 };
 
+export const cancelSaleAfterAcceptedByProvider = id => async (dispatch, getState, sdk) => {
+  if (cancelOrInProgress(getState())) {
+    return Promise.reject(new Error('Cancel already in progress'));
+  }
+  dispatch(cancelSaleRequest());
+
+  const currentTransactions = await sdk.transactions.show({ id });
+  const transType = currentTransactions.data.data.attributes.lastTransition === TRANSITION_ACCEPT
+    ? TRANSITION_PROVIDER_CANCEL_BEFORE_EXPIRE
+    : TRANSITION_PROVIDER_CANCEL_AFTER_EXPIRE;
+
+  return sdk.transactions
+    .transition({ id, transition: transType, params: {} }, { expand: true })
+    .then(response => {
+      dispatch(addMarketplaceEntities(response));
+      dispatch(cancelSaleSuccess());
+      dispatch(fetchExpireCurrentUserNotifications());
+      return response;
+    })
+    .catch(e => {
+      dispatch(cancelSaleError(storableError(e)));
+      log.error(e, 'provider-cancel-sale-failed', {
+        txId: id,
+        transition: transType,
+      });
+      throw e;
+    });
+};
+
+
```

Comment:
`cancelSaleAfterAcceptedByProvider`, `cancelSaleAfterAcceptedByCustomer`, `cancelSale `,... look similar in terms of logic can you generalise it a bit?